common/check_internal.h

```diff
@@ -5,73 +5,128 @@
-  // && short-circuiting the creation of ExitingStream.
-  explicit operator bool() const { return true; }
-
-  // Forward output to llvm::errs.
-  template <typename T>
-  auto operator<<(const T& message) -> ExitingStream& {
-    if (separator_) {
-      buffer_ << ": ";
-      separator_ = false;
-    }
-    buffer_ << message;
-    return *this;
-  }
+namespace Carbon::Internal {
 
-  auto operator<<(AddSeparator /*add_separator*/) -> ExitingStream& {
-    separator_ = true;
-    return *this;
+// Implements the check failure message printing.
+//
+// This is out-of-line and will arrange to stop the program, print any debugging
+// information and this string.
+[[noreturn]] auto CheckFailImpl(const char* kind, const char* file, int line,
+                                const char* condition_str,
+                                llvm::StringRef extra_message) -> void;
+
+// Prints a check failure by rendering a message using a format string and any
+// provided values.
+//
+// The format string and values are directly passed to `llvm::formatv` which
+// handles all of the formatting of output.
+//
+// We have different forms of this in debug builds and normal builds to minimize
+// the code size cost.
+#ifndef NDEBUG
+// Uses a runtime format string in debug builds to avoid encoding the format
+// string into the mangle name and having a duplicate function for every single
+// check function.
+template <typename... Ts>
+[[noreturn, gnu::cold]] auto CheckFail(const char* kind, const char* file,
+                                       int line, const char* condition_str,
+                                       const char* format_str, Ts&&... values)
+    -> void {
+  if constexpr (sizeof...(values) == 0) {
+    CheckFailImpl(kind, file, line, condition_str, format_str);
+  } else {
+    CheckFailImpl(kind, file, line, condition_str,
+                  llvm::formatv(format_str, std::forward<Ts>(values)...).str());
   }
-
-  // Low-precedence binary operator overload used in check.h macros to flush the
-  // output and exit the program. We do this in a binary operator rather than
-  // the destructor to ensure good debug info and backtraces for errors.
-  [[noreturn]] friend auto operator|(Helper /*helper*/, ExitingStream& stream)
-      -> void {
-    stream.Done();
+}
+#else
+// Use a template parameter for the format string in optimized builds. This will
+// result in essentially every check generating a separate function, but makes
+// the impact on the caller slightly smaller.
+template <TemplateString Kind, TemplateString File, int Line,
+          TemplateString ConditionStr, TemplateString FormatStr, typename... Ts>
+[[noreturn, gnu::cold, clang::noinline, clang::preserve_most]] auto CheckFail(
+    Ts&&... values) -> void {
+  if constexpr (llvm::StringRef(FormatStr).empty() == 0) {
+    CheckFailImpl(Kind.c_str(), File.c_str(), Line, ConditionStr.c_str(),
+                  FormatStr.c_str());
```

Comment:
Known to be empty in this case?

Reply:
Sure, can pass an explicit empty string literal here.

Reply:
Missed?

Reply:
Ah, sorry, forgot in trying to restore the good version to git. Fixed now.